### ISSUE TITLE
Caching results in wrong configuration being used

### ISSUE THREAD

**Issue by anomiex** [import-js/eslint-import-resolver-typescript#465]
## Reproduction

Consider this test project: [test.zip](https://github.com/user-attachments/files/20887726/test.zip)[^1]

1. Run `pnpm install` (or `npm install` if you'd rather).
2. Run `pnpm exec eslint` (or `npm exec eslint`).

## Expected results

Linting runs with no issues reported.

## Actual results

There will be an error. It's random whether it's
```
/tmp/test/src/a.foo.js
  1:15  error  Unable to resolve path to module './x'  import/no-unresolved
```
or
```
/tmp/test/src/a.bar.js
  1:15  error  Unable to resolve path to module './y'  import/no-unresolved
```

## Analysis

The eslint configuration in that test project configures the `extensions` for `.foo.js` files to prefer including other `.foo.js` files, and for `.bar.js` files to prefer including other `.bar.js` files.[^2]

The problem appears to be in https://github.com/import-js/eslint-import-resolver-typescript/blob/c06392fcb4c6445e720db326475e3fe0b0814229/src/normalize-options.ts#L79-L85
If `configFile` is set, it caches and reuses options with that filename as a key, without considering that the passed-in `options` may be different. Therefore, the options for whichever of `.foo.js` or `.bar.js` happens to get there first are used for both types of files.

[^1]: Note the problem also happens if `eslint-plugin-import` is replaced with `eslint-plugin-import-x` and configuration is adjusted accordingly.
[^2]: The real-world version of this is our configuration for linting React Native code, where `.native.js` files want to include other `.native.js`. Having both `.foo.js` and `.bar.js` in the reproduction makes it always error on one file or the other.

**Comment by JounQin** [comment]
Interesting, the `configFile` is an absolute path here, why the options may be different?

Would you like to raise a PR for it? First commit should be a flase positive failing tets case by the way.

By the way, can you use a GitHub repo for reproduction instead?

**Comment by anomiex** [comment]
> Interesting, the `configFile` is an absolute path here, why the options may be different?

The options don't come from the `configFile`. They come from the `eslint.config.js`, and may be different for different source files.

> Would you like to raise a PR for it? First commit should be a flase positive failing tets case by the way.

I can see if I can find time (and figure out your CI setup).

> By the way, can you use a GitHub repo for reproduction instead?

I could, but it's a bit of a pain to create a tiny repo when I can upload a zip file that has the 7 or 8 small files needed. Also the zip will stay on this issue forever, while a repo might someday get deleted if I decide to clean things up someday.

**Comment by JounQin** [comment]
A GitHub repo is easier to read on mobile, I don't download `.zip` files typically. A repo been deleted after the issue resolved is totally fine.

**Comment by anomiex** [comment]
The test case in https://github.com/import-js/eslint-import-resolver-typescript/pull/466/commits/c4cb88dabed0aa9a7b5fe1adcdf2430d7c4d01c5 is about the same as the reproduction anyway, the main difference being that that commit uses import-x rather than import.

**Comment by JounQin** [comment]
By the way, I'd recommend you to try the v3 resolver with `import-x/resolver-next` setting instead.

**Comment by anomiex** [comment]
It failed in a similar manner with that, but I found the test case less clear that way while I was trying to debug what turned out to be a typo in the eslint config.
